src/embeddings/tools/sgns.cpp

```diff
@@ -0,0 +1,607 @@
+                            = noise_dist_[(buffer.next_random(buffer.engine)
+                                           >> 16)
+                                          % noise_dist_.size()];
+                        if (target == window[target_widx])
+                            continue;
+                        update_vectors(0, target, buffer.neu1e, l1);
+                    }
+
+                    for (sgns_net_vector::size_type i = 0; i < vector_size_;
+                         i++)
+                    {
+                        net_.syn0[l1 + i] += buffer.neu1e[i];
+                    }
+                }
+            }
+
+            // Load the next word in the document into the window.
+            if (add_next_index(window, buffer.stream_, word_counter,
+                               buffer.engine, buffer.next_random))
+            {
+                ++target_widx;
+
+                // If the window is now too big, drop the first element.
+                if (window.size() > (max_window_size_ * 2 + 1))
+                {
+                    window.pop_front();
+                    --target_widx;
+                }
+            }
+            else
+            {
+                // If the document is out of words, continue dropping
+                // elements off the front of the window.
+                window.pop_front();
+            }
+
+            // Update our learning rate every 10,000 words read.
+            if (word_counter >= 10000)
+            {
+                update_progress(word_counter, io_mutex, progress);
+                word_counter = 0;
+            }
+        }
+    }
+
+    // Update the vectors
+    void update_vectors(int label, sgns_net_vector::size_type& target,
+                        sgns_net_vector& neu1e,
+                        const sgns_net_vector::size_type& l1)
+    {
+        const sgns_net_vector::size_type l2 = target * vector_size_;
+
+        float dot_product = 0.0f;
+        for (sgns_net_vector::size_type i = 0; i < vector_size_; i++)
+        {
+            dot_product += net_.syn0[l1 + i] * net_.syn1neg[l2 + i];
+        }
+
+        const float sigma = 1.0f / (1.0f + (float)exp(-dot_product));
+        const float update = learning_rate_ * (label - sigma);
+
+        for (sgns_net_vector::size_type i = 0; i < vector_size_; i++)
+        {
+            neu1e[i] += update * net_.syn1neg[l2 + i];
+            net_.syn1neg[l2 + i] += update * net_.syn0[l1 + i];
+        }
+    }
+
+    // Update the progress indicator
+    void update_progress(int64_t word_counter, std::mutex& io_mutex,
+                         printing::progress& progress)
+    {
+        word_count_actual_ += word_counter;
+
+        learning_rate_ = std::max(
+            starting_learning_rate_
+                * (1.0f
+                   - (float)word_count_actual_
+                         / (float)(iterations_ * vocab_.total_count + 1.0f)),
+            starting_learning_rate_ * 0.0001f);
```

Comment:
This update is a race condition. I think it also is in the original word2vec, but it's perhaps worth noting.